### ISSUE TITLE
`Token.position.end.line` super slow?

### ISSUE THREAD

**Issue by olafurpg** [scalameta/scalameta#334]
While profiling `scalafmt` I was surprised to find out that an innocent call to `Token.position.end.line` triggered a call to `scala.meta.inputs.Point$Offset.x$1$lzycompute` which occupied 10% of precious CPU time, see screenshot:

<img width="829" alt="screen shot 2016-02-20 at 11 31 19" src="https://cloud.githubusercontent.com/assets/1408093/13196111/bc1c2532-d7c5-11e5-9143-2050e93212de.png">

Any idea what might be the bottleneck here? I use that method to format only a subset of lines instead of the whole file.

**Comment by xeno-by** [comment]
Looks like it's very slow, because it scans the entire file: https://github.com/xeno-by/scalameta/blob/master/scalameta/tokens/src/main/scala/scala/meta/inputs/Position.scala#L54. Any idea how we could improve here?

**Comment by xeno-by** [comment]
Allright, I think we could probably precache this info somehow while parsing. How important is this for you?

**Comment by olafurpg** [comment]
Not important at all for now at least. I only use it for formatting a subset of lines, and that feature is half-broken anyways. In the end, I might even implement the feature without using that method.

**Comment by xeno-by** [comment]
Should be fixed now in `0.1.0-SNAPSHOT`. Could you verify please?

**Comment by olafurpg** [comment]
I can confirm the speed improvements are **significant**, using `token.position.end.line` on a hot path in scalafmt:

```
0.1.0-SNAPSHOT
[info] Benchmark           Mode  Cnt     Score    Error  Units
[info] GenJsCode.scalafmt  avgt   10  1520.984 ± 92.267  ms/op
0.1.0-RC1
[info] Benchmark           Mode  Cnt      Score     Error  Units
[info] GenJsCode.scalafmt  avgt   10  13500.775 ± 233.179  ms/op
```
That's more than 10x 👍 👍 👍 This will be vital for incremental formatting in scalafmt, thank!
